feat: retain cycles in kotlin (#4402)
JetBrains/kotlin-web-site: docs/topics/native/native-ios-integration.md

[//]: # (title: Integration with Swift/Objective-C ARC)

> Kotlin and Objective-C use different memory management strategies, Kotlin has a tracing garbage collector,

Reviewer: ```suggestion
Kotlin and Objective-C use different memory management strategies. Kotlin has a tracing garbage collector,
```

> Kotlin and Objective-C have two different memory models, Kotlin uses a garbage collector,
> while Objective-C relies on automatic reference counting (ARC).

Reviewer 2: ```suggestion
Kotlin and Objective-C use different memory management strategies, Kotlin uses a tracing garbage collector,
while Objective-C relies on automatic reference counting (ARC).
```
How about this?

Don't talk about whole memory models, just about memory management strategies. And also specify that Kotlin uses a *tracing* garbage collector (because ARC is also a form of garbage collection, but without tracing)

> The integration between them is usually seamless and generally requires no additional work.

Reviewer: Being a new paragraph, it's worth to clarify what we refer with "them" and mention the nouns again.

Author: I think mentioning nouns would bring too much repetition, but I clarified the pronoun though

after the GC cycle. The limitation stems from `deinit`, which can call arbitrary code, including the Kotlin code that
cannot be run during the GC pause.

### Retain cycles

In a retain cycle, a number of objects refer each other using strong references cyclically:

![Retain cycles](native-retain-cycle.png){height=200}

> Kotlin's GC and Objective-C's ARC handle retain cycles differently. When objects become unreachable, the GC can properly
> reclaim such cycles, while Objective-C's ARC can't. Therefore, retain cycles of Kotlin objects can be reclaimed,
> while retain cycles of Swift/Objective-C objects cannot.

Reviewer 2: ```suggestion
Kotlin's tracing GC and Objective-C's ARC handle retain cycles differently. When objects become unreachable, tracing GC can properly
reclaim such cycles, while ARC can't. Therefore, retain cycles of Kotlin objects can be reclaimed,
while retain cycles of Swift/Objective-C objects cannot.
```

How about this? Once again, spell out *tracing* GC.

Reviewer 2: Maybe it also makes sense to link to https://docs.swift.org/swift-book/documentation/the-swift-programming-language/automaticreferencecounting/#Strong-Reference-Cycles-Between-Class-Instances ?

Consider the case when retain cycles contain both Objective-C and Kotlin objects:

![Retain cycles with Objective-C and Kotlin objects](native-objc-kotlin-retain-cycles.png){height=150}

This involves combining Kotlin's and Objective-C's memory management models that cannot handle (reclaim) retain cycles
together. That means that if at least one Objective-C object is present, the retain cycle of a whole graph of objects
cannot be reclaimed, and it's impossible to break the cycle from the Kotlin side.

> Unfortunately, no special instruments are currently available to detect retain cycles in Kotlin/Native code.

Reviewer 2: Maybe we can suggest to use the same techniques as in Swift to break retain cycles? For example, https://docs.swift.org/swift-book/documentation/the-swift-programming-language/automaticreferencecounting/#Weak-References or https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/mmPractical.html#//apple_ref/doc/uid/TP40004447-1000810

Author: Thank you! Both links describe weak references, I chose a more general section from the first article

after the GC cycle. The limitation stems from `deinit`, which can call arbitrary code, including the Kotlin code that
cannot be run during the GC pause.

### Retain cycles

In a _retain cycle_, a number of objects refer each other using strong references cyclically:

![Retain cycles](native-retain-cycle.png){height=200}

> Kotlin's tracing GC and Objective-C's ARC handle retain cycles differently. When objects become unreachable, tracing GC

Reviewer: ```suggestion
Kotlin's tracing GC and Objective-C's ARC handle retain cycles differently. When objects become unreachable, Kotlin tracing GC
```

can properly reclaim such cycles, while Objective-C's ARC cannot. Therefore, retain cycles of Kotlin objects can be reclaimed,
while [retain cycles of Swift/Objective-C objects cannot](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/automaticreferencecounting/#Strong-Reference-Cycles-Between-Class-Instances).

Consider the case when a retain cycle contains both Objective-C and Kotlin objects:

![Retain cycles with Objective-C and Kotlin objects](native-objc-kotlin-retain-cycles.png){height=150}

This involves combining Kotlin's and Objective-C's memory management models that cannot handle (reclaim) retain cycles

> together. That means that if at least one Objective-C object is present, the retain cycle of a whole graph of objects

Reviewer: ```suggestion
together. Meaning that if at least one Objective-C object is present, the retain cycle of a whole graph of objects
```

Reviewer: Or consider even removing that part and starting the sentence with "If". To avoid repeating "that" so closely.

Author: I removed the second 'that' to avoid repetition, I think that works too
